File: 백준/Silver/14501. 퇴사/퇴사.py
# https://www.acmicpc.net/problem/14501
# DP
import sys

N = int(sys.stdin.readline().rstrip())
counseling = []
for _ in range(N):
    time, pay = map(int, sys.stdin.readline().rstrip().split())
    counseling.append((time, pay))

dp = [0]*(N+1)

for i in range(N):
    t = counseling[i][0]
    p = counseling[i][1]
    if i+t <= N:       
        dp[i+t] = max(dp[i+t], dp[i]+p)
        for j in range(i+t+1, N+1):
            if dp[j] < dp[i+t]:
                dp[j] = dp[i+t]

print(dp[-1])

File: 백준/Silver/14501. 퇴사/퇴사.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int[] timeRequired = new int[n];
        int[] pay = new int[n];
        
        // 상담에 필요한 기간은 5일 이하이기 때문
        int[] dp = new int[n+5];

        for(int i = 0; i < n; i++){
            String[] inputs = br.readLine().split(" ");
            timeRequired[i] = Integer.parseInt(inputs[0]);
            pay[i] = Integer.parseInt(inputs[1]);
        }

        for(int i = 0; i < n; i++){
            for(int j = i+timeRequired[i]; j < n+5; j++){
                dp[j] = Math.max(dp[j], dp[i]+pay[i]);
//              System.out.println(j + " " + dp[j]);

            }
//            for(int k = 0; k < n+5; k++){
//                System.out.print(dp[k] + " ");
//            }
//            System.out.println();
        }

        System.out.println(dp[n]);
    }
}
